iocell/src/main/scala/chisel/IOCell.scala

```diff
@@ -0,0 +1,188 @@
+
+  def generateIOFromSignal[T <: Data](coreSignal: T, name: Option[String] = None,
+    inFn: () => DigitalInIOCell = IOCell.exampleInput,
+    outFn: () => DigitalOutIOCell = IOCell.exampleOutput,
+    anaFn: () => AnalogIOCell = IOCell.exampleAnalog): (T, Seq[IOCell]) =
+  {
+    val padSignal = IO(DataMirror.internal.chiselTypeClone[T](coreSignal))
+    val iocells = IOCell.generateFromSignal(coreSignal, padSignal, name, inFn, outFn, anaFn)
+    (padSignal, iocells)
+  }
+
+  def generateFromSignal[T <: Data](coreSignal: T, padSignal: T, name: Option[String] = None,
+    inFn: () => DigitalInIOCell = IOCell.exampleInput,
+    outFn: () => DigitalOutIOCell = IOCell.exampleOutput,
+    anaFn: () => AnalogIOCell = IOCell.exampleAnalog): Seq[IOCell] =
+  {
+    coreSignal match {
+      case coreSignal: Analog => {
+        if (coreSignal.getWidth == 0) {
+          Seq()
+        } else {
+          require(coreSignal.getWidth == 1, "Analogs wider than 1 bit are not supported because we can't bit-select Analogs (https://github.com/freechipsproject/chisel3/issues/536)")
+          val iocell = anaFn()
+          name.foreach(n => iocell.suggestName(n))
+          iocell.io.core <> coreSignal
+          padSignal <> iocell.io.pad
+          Seq(iocell)
+        }
+      }
+      case coreSignal: Clock => {
+        DataMirror.directionOf(coreSignal) match {
+          case ActualDirection.Input => {
+            val iocell = inFn()
+            name.foreach(n => iocell.suggestName(n))
+            coreSignal := iocell.io.i.asClock
+            iocell.io.ie := true.B
+            iocell.io.pad := padSignal.asUInt.asBool
+            Seq(iocell)
+          }
+          case ActualDirection.Output => {
+            val iocell = outFn()
+            name.foreach(n => iocell.suggestName(n))
+            iocell.io.o := coreSignal.asUInt.asBool
+            iocell.io.oe := true.B
+            padSignal := iocell.io.pad.asClock
+            Seq(iocell)
+          }
+          case _ => throw new Exception("Unknown direction")
+        }
+      }
+      case coreSignal: Bits => {
+        require(padSignal.getWidth == coreSignal.getWidth, "padSignal and coreSignal must be the same width")
+        if (padSignal.getWidth == 0) {
+          // This dummy assignment will prevent invalid firrtl from being emitted
+          DataMirror.directionOf(coreSignal) match {
+            case ActualDirection.Input => coreSignal := 0.U
+            case _ => {}
+          }
+          Seq()
+        } else {
+          DataMirror.directionOf(coreSignal) match {
+            case ActualDirection.Input => {
+              // this type cast is safe because we guarantee that padSignal and coreSignal are the same type (T), but the compiler is not smart enough to know that
+              val iocells = padSignal.asInstanceOf[Bits].asBools.zipWithIndex.map { case (w, i) =>
+                val iocell = inFn()
+                name.foreach(n => iocell.suggestName(n + "_" + i))
```

Comment:
It might be worth letting chisel do the deduplication of overlapping names here. That way it would match the normal chisel naming conventions.

Reply:
It can't; you get anonymous names.

Reply:
I meant `iocell.suggestName(n)`

Reply:
I'd be worried that it would not number them in the same order

Reply:
Ok I'm worried that we will have things like `myIO_0, myIO_1` whereas the core signal is `myIO, myIO_1`

Reply:
I see. hmm. Let's discuss offline. I'd blame Chisel for this as well :).

Reply:
Changed this as requested